### DISCUSSION TITLE
AssistantAgent: recognize generated answer from ChatMessageContent

### DISCUSSION THREAD

**Discussion by MaxiPigna** [microsoft/semantic-kernel#9730] [Q&A] [answered]
I have implemented an Assistant Agent with the Code Interpreter feature enabled. When I invoke the Agent, it provides multiple responses from a single invocation, which include both the intermediate reasoning steps and the generated answers for the user. I wish to filter out all intermediate steps and deliver only the generated answers to the user, because intermediate steps contain technical details like the python code executed by the agent. 
I noticed that **ChatMessageContent** contains a list of items, and each item can be of a different type. However, among the possible types (e.g., AnnotationContent, TextContent, FileReferenceContent, etc. ), I did not find the correct one to recognize them.

**Comment by crickman** [comment, marked as answer]
Thank you for this question @MaxiPigna.  

Content/reasoning produced by the code-interpreter includes a metadata flag so that it may be identified.  This key is published as the constant: `OpenAIAssistantAgent.CodeInterpreterMetadataKey`:

https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/InternalUtilities/samples/AgentUtilities/BaseAgentsTest.cs#L59

We initially considered marking these using the `Tool` role, but this creates issues due to specific semantics related to the usage this role for chat-completion.

Other customers have provided input they they very much want this generated code / internal reasoning to be exposed; although, I understand this will vary across different contexts.

**Reply by MaxiPigna** [reply]
It works, thanks! However, I noticed that I received multiple messages with the flag set to false, which didn't contain code instructions but rather reasoning statements. The solution was to override the final answer each time to ensure only the last message, which is the answer to the user, is taken.
